compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs

```diff
@@ -0,0 +1,192 @@
+              -> GenOutput
+extendGODelta delta = over goDel (concatDelta delta)
+
+genPackages :: MonadDelta m => [(PackageId, (Package, Maybe PackageName))]
+            -> m Delta
+genPackages inp = ask >>= \del0 -> foldM genPackage' del0 inp
+  where
+    -- TODO: These prime definitions are bit silly. Merge with the regular
+    -- functions.
+    -- TODO: This >>= return . concatDelta returns a lot. Figure out a way to
+    -- abstract over it. Same for this returning foldM.
+    genPackage' :: MonadDelta m => Delta
+                -> (PackageId, (Package, Maybe PackageName)) -> m Delta
+    genPackage' deli (idi, (paci, _)) =
+      -- TODO: This is getting quite unreadable.
+      foldM (\deli' modi -> genModule' deli' (PRImport idi, modi))
+            deli (NM.toList $ packageModules paci)
+      >>= return . (concatDelta deli)
+    genModule' :: MonadDelta m => Delta -> (PackageRef, Module) -> m Delta
+    genModule' deli (paci, modi) = introDelta deli (genModule paci modi)
+                                   >>= return . (concatDelta deli)
+
+-- TODO: Type synonyms and data types are ignored for now.
+genModule :: MonadDelta m => PackageRef -> Module -> m Delta
+genModule pac mod = do
+  del0 <- ask
+  del1 <- foldM genValue' del0 (NM.toList $ moduleValues mod)
+  del2 <- foldM genTemplate' del1 (NM.toList $ moduleTemplates mod)
+  return $ concatDelta del1 del2
+  where
+    genValue' :: MonadDelta m => Delta -> DefValue -> m Delta
+    genValue' deli vali = introDelta deli (genValue pac (moduleName mod) vali)
+                          >>= return . (concatDelta deli)
+    genTemplate' :: MonadDelta m => Delta -> Template -> m Delta
+    genTemplate' deli temi = introDelta deli (genTemplate temi)
+                          >>= return . (concatDelta deli)
+
+genValue :: MonadDelta m => PackageRef -> ModuleName -> DefValue -> m Delta
+genValue pac mod val = do
+  expOut <- genExpr (dvalBody val)
+  let qname = Qualified pac mod (fst $ dvalBinder val)
+  return $ set devals (singleton qname (expOut ^. goExp, expOut ^. goUpd)) emptyDelta
+
+-- TODO: Handle annotated choices, by returning a set of annotations.
+genChoice :: MonadDelta m => (Qualified TypeConName) -> TemplateChoice
+          -> m GenOutput
+genChoice tem cho = do
+  expOut <- extVarDelta (fst $ chcArgBinder cho) $ genExpr (chcUpdate cho)
+  let updSet = if chcConsuming cho
+        -- TODO: Convert the `ExprVarName`s to `FieldName`s
+        then over usArc ((:) (UpdArchive tem [])) (expOut ^. goUpd)
+        else expOut ^. goUpd
+  return $ set goUpd updSet
+         $ over (goDel . devars) ((:) (fst $ chcArgBinder cho))
+         expOut
+
+genTemplate :: MonadDelta m => Template -> m Delta
+genTemplate = undefined -- TODO
+
+genExpr :: MonadDelta m => Expr -> m GenOutput
+genExpr = \case
+  ETmApp fun arg  -> genForTmApp fun arg
+  ETyApp expr typ -> genForTyApp expr typ
+  EVar name       -> genForVar name
+  EVal w          -> genForVal w
```

Comment:
Please don't align the arrows in case expressions. This will lead to a lot of "diff noise" down the line.